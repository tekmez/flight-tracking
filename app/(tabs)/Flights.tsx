import {
  Platform,
  SafeAreaView,
  Text,
  TouchableOpacity,
  View,
} from "react-native";
import { StatusBar } from "expo-status-bar";
import { StatusBar as NativeBar } from "react-native";
import Feather from "@expo/vector-icons/Feather";
import EmptyState from "@/components/EmptyState";
import TicketCard from "@/components/TicketCard";
import { useState } from "react";
import { Flight } from "@/types/FlightType";
import TButton from "@/components/ui/Button";
import Modal from "@/components/Modal";
import TicketList from "@/components/TicketList";
import {
  airlines,
  cities,
  dates,
  durations,
  times,
} from "@/constants/AirlineConstants";

const Flights = () => {
  const [flights, setFlights] = useState<Flight[] | []>([]);
  const [selectedFlight, setSelectedFlight] = useState<Flight | null>(null);

  const getRandomElement = (array: string[]) =>
    array[Math.floor(Math.random() * array.length)];

  const generateRandomFlight = () => {
    const airline = getRandomElement(airlines);
    const departureCity = getRandomElement(cities);
    const arrivalCity = getRandomElement(
      cities.filter((city) => city !== departureCity)
    );
    const takeOffDate = getRandomElement(dates);
    const landingDate = getRandomElement(
      dates.filter((date) => date !== takeOffDate)
    );
    const flightNumber = `${airline.split(" ")[0].charAt(0)}${airline
      .split(" ")[0]
      .charAt(3)
      .toUpperCase()} ${Math.floor(Math.random() * 10000)}`;
    return {
      flightNumber,
      airline,
      departureCity,
      arrivalCity,
      departureTime: getRandomElement(times),
      arrivalTime: getRandomElement(times),
      duration: getRandomElement(durations),
      takeOffDate,
      landingDate,
    };
  };

  const handleAddFlight = () => {
    setFlights([...flights, generateRandomFlight()]);
  };

  const handleOpenModal = (flight: Flight) => {
    setSelectedFlight(flight);
  };

  const handleCloseModal = () => {
    setSelectedFlight(null);
  };
  const handleRemoveFlight = (flight: Flight) => {
    setFlights(flights.filter((f) => f.flightNumber !== flight.flightNumber));
    setSelectedFlight(null);
  };
  return (
    <SafeAreaView
      className="flex-1 bg-white"
      style={{
        paddingTop: Platform.OS === "android" ? NativeBar.currentHeight : 0,
      }}
    >
      <StatusBar style="dark" />
      <View className="flex-row items-baseline justify-between py-1 px-4">
        <Text className="font-sfProBold text-[34px] leading-[41px] text-gray-900">
          Flights
        </Text>
        <TouchableOpacity onPress={handleAddFlight}>
          <Feather
            name="plus"
            size={24}
            color="#111827"
            testID="add-flight-button"
          />
        </TouchableOpacity>
      </View>
      <View className="mt-3 flex-1 px-4">
        <TicketList
          flights={flights}
          handleOpenModal={handleOpenModal}
          handleAddFlight={handleAddFlight}
          EmptyState={EmptyState}
          testID="ticket-list"
        />
      </View>
      <Modal visible={selectedFlight !== null} onClose={handleCloseModal}>
        {selectedFlight && <TicketCard {...selectedFlight} />}
        <TButton
          onPress={() => handleRemoveFlight(selectedFlight as Flight)}
          title="Remove Flight"
          classNames="mt-4 bg-[#FEE2E2] w-full"
          textClassNames="text-red-900 font-sfProMedium text-center"
        />
      </Modal>
    </SafeAreaView>
  );
};

export default Flights;
