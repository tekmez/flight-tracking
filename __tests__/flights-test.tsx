import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import Flights from "@/app/(tabs)/Flights";

jest.mock("@expo/vector-icons/Feather", () => "Feather");
jest.mock("@/components/EmptyState", () => "EmptyState");
jest.mock("@/components/TicketCard", () => "TicketCard");
jest.mock("@/components/Modal", () => "Modal");
jest.mock("@/components/TicketList", () => "TicketList");

describe("Flights", () => {
  it("renders correctly", () => {
    const tree = render(<Flights />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("adds a flight when the plus button is pressed", () => {
    const { getByTestId } = render(<Flights />);
    const addButton = getByTestId("add-flight-button");

    fireEvent.press(addButton);

    expect(getByTestId("ticket-list")).toBeTruthy();
  });

  it("opens modal when a flight is selected", () => {
    const { getByTestId, getByText } = render(<Flights />);
    const addButton = getByTestId("add-flight-button");

    fireEvent.press(addButton);

    const ticketList = getByTestId("ticket-list");
    fireEvent(ticketList, "handleOpenModal", { flightNumber: "TEST123" });

    expect(getByTestId("ticket-list")).toBeTruthy();
    expect(getByText("Remove Flight")).toBeTruthy();
  });

  it("removes a flight when remove button is pressed", () => {
    const { getByTestId, getByText, queryByText } = render(<Flights />);
    const addButton = getByTestId("add-flight-button");

    fireEvent.press(addButton);

    const ticketList = getByTestId("ticket-list");
    fireEvent(ticketList, "handleOpenModal", { flightNumber: "TEST123" });

    const removeButton = getByText("Remove Flight");
    fireEvent.press(removeButton);

    expect(queryByText("TEST123")).toBeNull();
  });
});
