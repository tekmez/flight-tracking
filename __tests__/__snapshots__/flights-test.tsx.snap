// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flights renders correctly 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "backgroundColor": "#fff",
      },
      {
        "paddingTop": 0,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "alignItems": "baseline",
        },
        {
          "justifyContent": "space-between",
        },
        {
          "paddingBottom": 4,
          "paddingTop": 4,
        },
        {
          "paddingLeft": 16,
          "paddingRight": 16,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "SFProBold",
          },
          {
            "fontSize": 34,
          },
          {
            "lineHeight": 41,
          },
          {
            "color": "#111827",
          },
        ]
      }
    >
      Flights
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Feather
        color="#111827"
        name="plus"
        size={24}
        testID="add-flight-button"
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "marginTop": 12,
        },
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "paddingLeft": 16,
          "paddingRight": 16,
        },
      ]
    }
  >
    <TicketList
      EmptyState="EmptyState"
      flights={[]}
      handleAddFlight={[Function]}
      handleOpenModal={[Function]}
      testID="ticket-list"
    />
  </View>
  <Modal
    onClose={[Function]}
    visible={false}
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#FEE2E2",
          "borderBottomLeftRadius": 14,
          "borderBottomRightRadius": 14,
          "borderTopLeftRadius": 14,
          "borderTopRightRadius": 14,
          "marginTop": 16,
          "opacity": 1,
          "paddingBottom": 14,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 14,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "SFProRegular",
            },
            {
              "color": "#fff",
            },
            {
              "fontSize": 17,
            },
            {
              "lineHeight": 22,
            },
            {
              "color": "#7f1d1d",
            },
            {
              "fontFamily": "SFProMedium",
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        Remove Flight
      </Text>
    </View>
  </Modal>
</RCTSafeAreaView>
`;
